Project/DAQ_System/Program/Src/Sensor/GyroScope/LSM6DSOXGyroscope.cpp

```diff
@@ -0,0 +1,169 @@
+        sum[2] += temp_data[2];
+    }
+
+    offset_average_[0] = sum[0] / average_sample_size;
+    offset_average_[1] = sum[1] / average_sample_size;
+    offset_average_[2] = sum[2] / average_sample_size;
+}
+
+void LSM6DSOX_Gyroscope::SetODR(ODR SelectedODRValue){
+
+		static constexpr uint8_t kNumBytes = 2;
+	    uint8_t commands[kNumBytes] = {0};
+	    commands[0] = LSM6DSOX_CTRL2_G;
+
+	    // used to store the original register values
+	    uint8_t* currentRegisterValue = 0;
+	    // contains the value of ODR to be written to register
+
+	    // read the current register values for the CTRL_X register
+		HAL_I2C_Master_Transmit(&i2c_,LSM6DSOX_GYR_ADDRESS, commands,1, HAL_MAX_DELAY);
+		HAL_I2C_Master_Receive(&i2c_,LSM6DSOX_GYR_ADDRESS, currentRegisterValue,1, HAL_MAX_DELAY);
+
+		// set the FSR bits to the desired value without touching the other bits in the register
+		commands[1] = (currentRegisterValue[0] & 0x0f) | static_cast<uint8_t>(SelectedODRValue) << 4 ;
+
+		// write to register with the desired bit values
+		HAL_I2C_Master_Transmit(&i2c_,LSM6DSOX_GYR_ADDRESS, commands,kNumBytes, HAL_MAX_DELAY);
+}
+
+void LSM6DSOX_Gyroscope::SetFSR(FSR SelectedFSRValue){
+
+	static constexpr uint8_t kNumBytes = 2;
+    uint8_t commands[kNumBytes] = {0};
+    commands[0] = LSM6DSOX_CTRL2_G;
+
+    //  used to store the original register values
+    uint8_t currentRegisterValue[1] = {0};
+
+    // read the current register values for the CTRL_X register
+
+	HAL_I2C_Master_Transmit(&i2c_,LSM6DSOX_GYR_ADDRESS, commands,1, HAL_MAX_DELAY);
+	HAL_I2C_Master_Receive(&i2c_,LSM6DSOX_GYR_ADDRESS, currentRegisterValue,1, HAL_MAX_DELAY);
+
+	// set the FSR bits to the desired value without touching the other bits in the register
+	commands[1] = (currentRegisterValue[0] & 0xF3) | (static_cast<uint8_t>(SelectedFSRValue) << 2 );
+
+	// write to register with the desired bit values
+	HAL_I2C_Master_Transmit(&i2c_,LSM6DSOX_GYR_ADDRESS, commands,kNumBytes, HAL_MAX_DELAY);
+
+	// setting up sensitivity factor based on the selected FSR Value. used to convert raw values to degrees
+	switch(SelectedFSRValue){
+	case FSR::DPS250:
+		SensitivityFactor = 0.00875;
+		break;
+	case FSR::DPS500:
+		SensitivityFactor = 0.01750;
+		break;
+	case FSR::DPS1000:
+		SensitivityFactor =  0.0350;
+		break;
+	case FSR::DPS2000:
+		SensitivityFactor = 0.0700;
+		break;
+
+	}
+
+}
+
+
+bool LSM6DSOX_Gyroscope::WriteReg(uint8_t addr_i2c, uint8_t addr_reg, uint8_t register_data){
+    uint8_t data[2] = {addr_reg, register_data};
+    return HAL_I2C_Master_Transmit(&i2c_,addr_i2c, data,2, HAL_MAX_DELAY);
+}
+
+
+
+bool LSM6DSOX_Gyroscope::Receive(uint8_t data_buf[]){
+	uint8_t addr_reg = LSM6DSOX_OUTX_L_G;
+	if(HAL_I2C_Master_Transmit(&i2c_, LSM6DSOX_GYR_ADDRESS, &addr_reg, 1, HAL_MAX_DELAY) != HAL_OK){
+		return false;
+	}
+	if(HAL_I2C_Master_Receive(&i2c_, LSM6DSOX_GYR_ADDRESS, data_buf, 6, HAL_MAX_DELAY) != HAL_OK){
+		return false;
+	}
+
+	return true;
+}
```

Comment:
I think we should tear out this `Receive()` function. The `LSM6DSOX_OUTX_L_G` is hard-coded, so you may be better off simply using `HAL_I2C_Master_Transmit()` and `HAL_I2C_Master_Receive()` directly.

(This function is called inside of `ReadRawData()`).

These will then be simplified with the new `II2C` interface at some point (which is outside the scope of this task).

Reply:
sure I can do that